drivers/at86rf2xx/at86rf2xx.c

```diff
@@ -32,59 +35,215 @@
+        params++;
+    }
 #else
-    /* initialize device descriptor */
-    dev->params = *params;
+    (void)devs;
+    (void)params;
+    (void)num;
 #endif
 }
 
-static void at86rf2xx_disable_clock_output(at86rf2xx_t *dev)
+void at86rf231_setup(at86rf231_t *devs, const at86rf231_params_t *params,
+                     uint8_t num)
 {
-#if defined(MODULE_AT86RFA1) || defined(MODULE_AT86RFR2)
-    (void) dev;
+#if IS_USED(MODULE_AT86RF231)
+    for (int i = 0; i < num; i++) {
+        devs->base.dev_type = AT86RF2XX_DEV_TYPE_AT86RF231;
+        devs->params = *params;
+        at86rf2xx_setup((at86rf2xx_t *)devs);
+        devs++;
+        params++;
+    }
 #else
-    uint8_t tmp = at86rf2xx_reg_read(dev, AT86RF2XX_REG__TRX_CTRL_0);
-    tmp &= ~(AT86RF2XX_TRX_CTRL_0_MASK__CLKM_CTRL);
-    tmp &= ~(AT86RF2XX_TRX_CTRL_0_MASK__CLKM_SHA_SEL);
-    tmp |= (AT86RF2XX_TRX_CTRL_0_CLKM_CTRL__OFF);
-    at86rf2xx_reg_write(dev, AT86RF2XX_REG__TRX_CTRL_0, tmp);
+    (void)devs;
+    (void)params;
+    (void)num;
 #endif
 }
 
-static void at86rf2xx_enable_smart_idle(at86rf2xx_t *dev)
+void at86rf232_setup(at86rf232_t *devs, const at86rf232_params_t *params,
+                     uint8_t num)
 {
-#if AT86RF2XX_SMART_IDLE_LISTENING
-    uint8_t tmp = at86rf2xx_reg_read(dev, AT86RF2XX_REG__TRX_RPC);
-    tmp |= (AT86RF2XX_TRX_RPC_MASK__RX_RPC_EN |
-            AT86RF2XX_TRX_RPC_MASK__PDT_RPC_EN |
-            AT86RF2XX_TRX_RPC_MASK__PLL_RPC_EN |
-            AT86RF2XX_TRX_RPC_MASK__XAH_TX_RPC_EN |
-            AT86RF2XX_TRX_RPC_MASK__IPAN_RPC_EN);
-    at86rf2xx_reg_write(dev, AT86RF2XX_REG__TRX_RPC, tmp);
-    at86rf2xx_set_rxsensitivity(dev, RSSI_BASE_VAL);
+#if IS_USED(MODULE_AT86RF232)
+    for (int i = 0; i < num; i++) {
+        devs->base.dev_type = AT86RF2XX_DEV_TYPE_AT86RF232;
+        devs->params = *params;
+        at86rf2xx_setup((at86rf2xx_t *)devs);
+        devs++;
+        params++;
+    }
 #else
-    (void) dev;
+    (void)devs;
+    (void)params;
+    (void)num;
+#endif
+}
+
+void at86rf233_setup(at86rf233_t *devs, const at86rf233_params_t *params,
+                     uint8_t num)
+{
+#if IS_USED(MODULE_AT86RF233)
+    for (int i = 0; i < num; i++) {
+        devs->base.dev_type = AT86RF2XX_DEV_TYPE_AT86RF233;
+        devs->params = *params;
+        at86rf2xx_setup((at86rf2xx_t *)devs);
+        devs++;
+        params++;
+    }
+#else
+    (void)devs;
+    (void)params;
+    (void)num;
+#endif
+}
+
+void at86rfa1_setup(at86rfa1_t *dev)
+{
+#if IS_USED(MODULE_AT86RFA1)
+    dev->base.dev_type = AT86RF2XX_DEV_TYPE_AT86RFA1;
+    at86rf2xx_setup((at86rf2xx_t *)dev);
+#else
+    (void)dev;
+#endif
+}
+
+void at86rfr2_setup(at86rfr2_t *dev)
+{
+#if IS_USED(MODULE_AT86RFR2)
+    dev->base.dev_type = AT86RF2XX_DEV_TYPE_AT86RFR2;
+    at86rf2xx_setup((at86rf2xx_t *)dev);
+#else
+    (void)dev;
+#endif
+}
+
+size_t at86rf2xx_get_size(const at86rf2xx_t *dev)
+{
+    switch (dev->base.dev_type) {
+        default: return sizeof(at86rf2xx_t);
```

Comment:
```suggestion
        default:
```

This would allow the compiler to optimize this function much better if only a single type of AT86RF2xx devices is used: The default would just fall through to the single remaining `case` if only one device is used, so that all checks can be dropped in that case.

Reply:
we could actually drop this function if there are variant specific functions (see https://github.com/RIOT-OS/RIOT/pull/12914/commits/98797e0ae3b756237541e25915106da9bb9c6176#r356007674)

Reply:
Ok

Reply:
> we could actually drop this function if there are variant specific functions (see [98797e0#r356007674](https://github.com/RIOT-OS/RIOT/commit/98797e0ae3b756237541e25915106da9bb9c6176#r356007674))

You mean get rid of `at86rf2xx_get_size()` and add `at86rf212b_get_size()` and so on?
This function is used to iterate over all devices if you only have an `at86rf2xx_t *`.
I think you must know the device type in any case before you call either `at86rf2xx_get_size(dev)` or e.g. `at86rf212b_get_size()`.